scapy/contrib/automotive/enumerator.py

```diff
@@ -0,0 +1,316 @@
+# Copyright (C) Nils Weiss <nils@we155.de>
+# This program is published under a GPLv2 license
+
+# scapy.contrib.description = Enumerator and Automotive Scanner Baseclasses
+# scapy.contrib.status = loads
+
+from collections import defaultdict, namedtuple
+from statistics import mean
+
+from scapy.error import Scapy_Exception, log_interactive
+from scapy.utils import make_lined_table
+from scapy.modules import six
+from scapy.contrib.automotive.ecu import ECU_State
+
+
+class Graph:
+    def __init__(self):
+        """
+        self.edges is a dict of all possible next nodes
+        e.g. {'X': ['A', 'B', 'C', 'E'], ...}
+        self.weights has all the weights between two nodes,
+        with the two nodes as a tuple as the key
+        e.g. {('X', 'A'): 7, ('X', 'B'): 2, ...}
+        """
+        self.edges = defaultdict(list)
+        self.weights = {}
+
+    def add_edge(self, from_node, to_node, weight=1):
+        # Note: assumes edges are bi-directional
+        self.edges[from_node].append(to_node)
+        self.edges[to_node].append(from_node)
+        self.weights[(from_node, to_node)] = weight
+        self.weights[(to_node, from_node)] = weight
+
+    @property
+    def nodes(self):
+        return self.edges.keys()
+
+    @staticmethod
+    def dijsktra(graph, initial, end):
+        # shortest paths is a dict of nodes
+        # whose value is a tuple of (previous node, weight)
+        shortest_paths = {initial: (None, 0)}
+        current_node = initial
+        visited = set()
+
+        while current_node != end:
+            visited.add(current_node)
+            destinations = graph.edges[current_node]
+            weight_to_current_node = shortest_paths[current_node][1]
+
+            for next_node in destinations:
+                weight = \
+                    graph.weights[(current_node, next_node)] + \
+                    weight_to_current_node
+                if next_node not in shortest_paths:
+                    shortest_paths[next_node] = (current_node, weight)
+                else:
+                    current_shortest_weight = shortest_paths[next_node][1]
+                    if current_shortest_weight > weight:
+                        shortest_paths[next_node] = (current_node, weight)
+
+            next_destinations = {node: shortest_paths[node] for node in
+                                 shortest_paths if node not in visited}
+            if not next_destinations:
+                return None
+            # next node is the destination with the lowest weight
+            current_node = min(next_destinations,
+                               key=lambda k: next_destinations[k][1])
+
+        # Work back through destinations in shortest path
+        path = []
+        while current_node is not None:
+            path.append(current_node)
+            next_node = shortest_paths[current_node][0]
+            current_node = next_node
+        # Reverse path
+        path = path[::-1]
```

Comment:
You can use `.reverse()` here, which is faster:

```
In [29]: %%time 
    ...: for _ in range(0, 100000): 
    ...:     a.reverse() 
    ...:                                                                                                                                                                           
CPU times: user 40.6 ms, sys: 36 µs, total: 40.7 ms
Wall time: 39.6 ms

In [30]: %%time 
    ...: for _ in range(0, 100000): 
    ...:     a = a[::-1] 
    ...:                                                                                                                                                                           
CPU times: user 96.3 ms, sys: 146 µs, total: 96.5 ms
Wall time: 95.3 ms
```

Reply:
Thanks